src/rust/ide/json-rpc/src/macros.rs

```diff
@@ -0,0 +1,129 @@
+        #[derive(Debug)]
+        pub struct Client {
+            /// JSON-RPC protocol handler.
+            handler : RefCell<Handler<Notification>>,
+        }
+
+        impl Client {
+                /// Create a new client that will use given transport.
+                pub fn new(transport:impl json_rpc::Transport + 'static) -> Self {
+                    let handler = RefCell::new(Handler::new(transport));
+                    Self { handler }
+                }
+
+                /// Asynchronous event stream with notification and errors.
+                ///
+                /// On a repeated call, previous stream is closed.
+                pub fn events(&self) -> impl Stream<Item = Event> {
+                    self.handler.borrow_mut().handler_event_stream()
+                }
+
+                /// Returns a future that performs any background, asynchronous work needed
+                /// for this Client to correctly work. Should be continually run while the
+                /// `Client` is used. Will end once `Client` is dropped.
+                pub fn runner(&self) -> impl Future<Output = ()> {
+                    self.handler.borrow_mut().runner()
+                }
+        }
+
+        impl API for Client {
+            $(fn $method(&self, $($param_name:$param_ty),*) -> std::pin::Pin<Box<dyn Future<Output=Result<$result>>>> {
+                let input = $method_input { $($param_name:$param_name),* };
+                Box::pin(self.handler.borrow().open_request(input))
+            })*
+        }
+
+        $(
+            /// Structure transporting method arguments.
+            #[derive(Serialize,Deserialize,Debug,PartialEq)]
+            #[serde(rename_all = "camelCase")]
+            struct $method_input {
+                $($param_name : $param_ty),*
+            }
+
+            impl json_rpc::RemoteMethodCall for $method_input {
+                const NAME:&'static str = $rpc_name;
+                type Returned = $result;
+            }
+        )*
+
+
+
+        // ==================
+        // === MockClient ===
+        // ==================
+
+        /// Mock used for tests.
+        #[derive(Debug,Default)]
+        pub struct MockClient {
+            $($method_result : RefCell<HashMap<($($param_ty),+),Result<$result>>>,)*
+        }
+
+        impl API for MockClient {
+            $(fn $method(&self $(,$param_name:$param_ty)+) -> std::pin::Pin<Box<dyn Future<Output=Result<$result>>>> {
+                let mut result = self.$method_result.borrow_mut();
+                let result     = result.remove(&($($param_name),+)).unwrap();
+                Box::pin(async move { result })
+            })*
+        }
+
+        impl MockClient {
+            $(/// Sets `$method`'s result to be returned when it is called.
+            pub fn $set_result(&self $(,$param_name:$param_ty)+, result:Result<$result>) {
+                self.$method_result.borrow_mut().insert(($($param_name),+),result);
+            })*
```

Comment:
in case you are doing something liek that `$( ... )*` and you are not doing it in single line, treat it as multiline `{ ... }` thing. Sothis code should be formatted as:

```rust
            $(
                /// Sets `$method`'s result to be returned when it is called.
                pub fn $set_result(&self $(,$param_name:$param_ty)+, result:Result<$result>) {
                    self.$method_result.borrow_mut().insert(($($param_name),+),result);
                }
            )*
```